src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java

```diff
@@ -231,14 +239,69 @@ private boolean isUnusedImport(String imprt) {
     private void processIdent(DetailAST ast) {
         final DetailAST parent = ast.getParent();
         final int parentType = parent.getType();
-        if (parentType != TokenTypes.DOT
+        if ((parentType != TokenTypes.DOT
             && parentType != TokenTypes.METHOD_DEF
             || parentType == TokenTypes.DOT
-                && ast.getNextSibling() != null) {
+                && ast.getNextSibling() != null)
+                && !isEnumDefinitionOrType(ast)
+                && !isAnnotation(ast)
+                && !isInnerInterfaceOrClass(parentType)) {
             referenced.add(ast.getText());
         }
     }
 
+    /**
+     * Checks whether the given parentType of the AST is an inner interface or inner class.
+     *
+     * @param parentType The token type
+     * @return true if the token type is INTERFACE_DEF or CLASS_DEF
+     */
+    private static boolean isInnerInterfaceOrClass(int parentType) {
+        return parentType == TokenTypes.INTERFACE_DEF
+                || parentType == TokenTypes.CLASS_DEF;
+    }
+
+    /**
+     * Checks whether the given AST is an enum.
+     *
+     * @param ast the IDENT node to process
+     * @return true if the AST is an enum
+     */
+    private boolean isEnumDefinitionOrType(DetailAST ast) {
+        boolean result = false;
+        if (ast.getParent().getType() == TokenTypes.ENUM_CONSTANT_DEF) {
+            result = true;
+        }
+        else if (ast.getParent().getType() == TokenTypes.ENUM_DEF) {
+            result = true;
+            enums.add(ast.getText());
+        }
+        else {
+            if (enums.contains(ast.getText())) {
+                result = true;
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Checks whether the given AST is an annotation.
+     *
+     * @param ast The IDENT node to process
+     * @return true if the AST is an annotation
+     */
+    private boolean isAnnotation(DetailAST ast) {
+        boolean result = false;
+        if (ast.getParent().getType() == TokenTypes.ANNOTATION_DEF) {
+            result = true;
+            annotations.add(ast.getText());
```

Comment:
I'm afraid it can not be solved this way. Consider 
```
public class InputUnusedImportsEnumAndAnnotation {

    @Map                 // <= First usage
    enum Set {
        List(0);
        Set(int arg){}
    }

    InputUnusedImportsEnumAndAnnotation() {
        Set x = Set.List;
    }

    @interface Map {}   // <= Declared here

}
```

Same for enums, classes and interfaces:
```java
public class InputUnusedImportsInnerClassInInterface {

    InputUnusedImportsInnerClassInInterface() {
        List x = null;  // < used here
    }

    interface List { // Declared here
        class Set {}
    }
}
```

There should be two passes: collect types first, then check for imports. And this will be more complicated for nested types:
```java
import java.util.List; // no violation here
import java.util.Set; // violation

public class InputUnusedImportsInnerClassInInterface {

    InputUnusedImportsInnerClassInInterface() {
        List x = null; // java.util.List
    }

    static class Nested {
        public Nested() {
            List x = null; // Nested.List
        }

        interface List {
            class Set {
            }
        }
    }
}
```